iphone/TitaniumKit/TitaniumKit/Sources/API/TiApp.m

```diff
@@ -1237,6 +1249,111 @@ - (KrollBridge *)krollBridge
+                                                          usingQueue:nil
+                                                       launchHandler:^(__kindof BGTask *_Nonnull task) {
+                                                         [registeredBackgroundTasks addObject:task];
+                                                         [self handleBGTask:task];
+                                                       }];
+  }
+}
+
+- (void)handleBGTask:(BGTask *)task
+{
+  NSString *notificationName = kTiBackgroundProcessNotification;
+  if ([task isKindOfClass:[BGAppRefreshTask class]]) {
+    // Fo refresh task submit it again
+    [self submitTaskForIdentifier:task.identifier];
+    notificationName = kTiBackgroundFetchNotification;
+  }
+  NSString *key = [NSString stringWithFormat:@"BgTask-%@", task.identifier];
+
+  [self tryToPostBackgroundModeNotification:[NSMutableDictionary dictionaryWithObjectsAndKeys:key, @"handlerId", nil]
+                       withNotificationName:notificationName];
+
+  task.expirationHandler = ^{
+    if ([task isKindOfClass:[BGProcessingTask class]]) {
+      // Fo processing task, if it is not completed in time then only submit it again.
+      [self submitTaskForIdentifier:task.identifier];
+    }
+    [task setTaskCompletedWithSuccess:false];
+    [registeredBackgroundTasks removeObject:task];
+  };
+}
+
+- (void)submitTaskForIdentifier:(NSString *)identifier
+{
+  NSDictionary *backgroundTask = [self backgroundTaskForIdentifier:identifier];
+  if (backgroundTask) {
+    [self submitBackgroundTask:backgroundTask];
+  }
+}
+
+- (void)submitBackgroundTask:(NSDictionary *)bgTask
+{
+  BGTaskRequest *taskRequest;
+  if ([bgTask[@"type"] isEqualToString:@"process"]) {
+    taskRequest = [[[BGProcessingTaskRequest alloc] initWithIdentifier:bgTask[@"identifier"]] autorelease];
+    ((BGProcessingTaskRequest *)taskRequest).requiresNetworkConnectivity = [TiUtils boolValue:bgTask[@"networkConnect"] def:NO];
+    ((BGProcessingTaskRequest *)taskRequest).requiresExternalPower = [TiUtils boolValue:bgTask[@"powerConnect"] def:NO];
+  } else {
+    taskRequest = [[[BGAppRefreshTaskRequest alloc] initWithIdentifier:bgTask[@"identifier"]] autorelease];
+  }
+  taskRequest.earliestBeginDate = bgTask[@"beginDate"];
+
+  [BGTaskScheduler.sharedScheduler submitTaskRequest:taskRequest error:nil];
+}
+
+- (void)backgroundTaskCompletedForIdentifier:(NSString *)identifier
+{
+  for (BGTask *task in registeredBackgroundTasks) {
+    if ([task.identifier isEqualToString:identifier]) {
+      [task setTaskCompletedWithSuccess:YES];
+      [registeredBackgroundTasks removeObject:task];
+      break;
+    }
+  }
+}
+
+- (NSDictionary *_Nullable)backgroundTaskForIdentifier:(NSString *)identifier
+{
+  NSDictionary *bgTask = nil;
+  for (NSDictionary *backgroundTask in backgroundTasks) {
+    if (backgroundTask[@"identifier"] == identifier) {
+      bgTask = backgroundTask;
```

Comment:
Simply return here, so you save the extra assignment. Return `nil` by default after the loop